File: app/src/main/java/com/newchar/woolhelper/state/BaseStateLayout.java
package com.newchar.woolhelper.state;

import androidx.collection.SparseArrayCompat;

/**
 * @author wenliqiang@100tal.com
 * date            2019-12-21
 * @since 当前版本描述，
 * @since 迭代版本描述
 */
public abstract class BaseStateLayout<T> implements IStateLayout<T> {

    protected T host;
    /**
     * 当前处于那种 state
     */
    protected int state;
    protected SparseArrayCompat<StateLayout> layoutArrays = new SparseArrayCompat<>(4);

    @Override
    public void attach(T attach) {
        host = attach;
    }

    @Override
    public T getHost() {
        return host;
    }

    public int getState() {
        return state;
    }

    public void putLayout(int state, StateLayout stateLayout) {
        layoutArrays.put(state, stateLayout);
    }

    public StateLayout getLayout(int state) {
        return layoutArrays.get(state);
    }

}

File: app/src/main/java/com/newchar/woolhelper/replacer/BaseStateLayout.kt
package com.newchar.woolhelper.replacer

import androidx.collection.SparseArrayCompat
import androidx.viewpager.widget.ViewPager.OnPageChangeListener

/**
 * @author wenliqiang@100tal.com
 * date            2019-12-21
 * @since 当前版本描述，
 * @since 迭代版本描述
 */
abstract class BaseStateLayout<T> : IStateLayout<T?> {

    var hostAA : T? = TODO()

    /**
     * 当前处于那种 state
     */
    var state = 0

    protected var layoutArrays = SparseArrayCompat<StateLayout>(4)
    protected var mStateChangeListener: OnPageChangeListener? = null

    override fun attach(attach: T?) {
        hostAA = attach
    }

    fun putLayout(state: Int, stateLayout: StateLayout) {
        layoutArrays.put(state, stateLayout)
    }

    fun getLayout(state: Int): StateLayout? {
        return layoutArrays[state]
    }

}